src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java

```diff
@@ -117,36 +185,167 @@ public void visitToken(DetailAST ast) {
+                    log(currentToken, MSG_KEY, previousLineWithOverloadMethod);
                 }
-                methodIndexMap.put(methodName, currentIndex);
-                methodLineNumberMap.put(methodName, currentToken.getLineNo());
             }
             currentToken = currentToken.getNextSibling();
         }
     }
 
+    /**
+     * Compare the modifiers on a given method with the {@code modifierGroups} and find the
+     * first match to create a unique identifier for the group and method name combination.
+     *
+     * @param currentToken assumed to be a method
+     * @return string that uniquely represents the {@code modifierGroup} and method name
+     */
+    private String extractMethodDescription(DetailAST currentToken) {
+        final DetailAST method = currentToken.findFirstToken(TokenTypes.IDENT);
+        final String methodName = method.getText();
+
+        String descriptor = DEFAULT_MODIFIER_GROUP_PATTERN + SEPARATOR + methodName;
+        final String methodModifiers = extractModifiersString(currentToken);
+        for (Pattern pattern : modifierGroups) {
+            final Matcher matcher = pattern.matcher(methodModifiers);
+            if (matcher.find()) {
+                // method belongs to the first overload group which matches its modifiers
+                descriptor = pattern.pattern() + SEPARATOR + methodName;
+                break;
+            }
+        }
+        return descriptor;
+    }
+
+    /**
+     * Extract the modifier(s) of a method definition into a string.
+     *
+     * @param token expected to be TokenType.METHOD_DEF .
+     * @return modifier string for {@code token} with 'package' scope added if needed
+     */
+    private static String extractModifiersString(final DetailAST token) {
+        final DetailAST modifiersAst = token.findFirstToken(TokenTypes.MODIFIERS);
+        final List<String> collectedModifiers = new LinkedList<>();
+        DetailAST modifier = modifiersAst.getFirstChild();
+        boolean foundScopeModifier = false;
+        while (modifier != null) {
+            // skipping annotations
+            if (modifier.getType() != TokenTypes.ANNOTATION) {
+                final String currModifierText = modifier.getText();
+                if (!foundScopeModifier) {
+                    // no need to test again if scope modifier already found
+                    foundScopeModifier = isScopeModifier(currModifierText);
+                }
+
+                collectedModifiers.add(currModifierText);
+            }
+
+            // advance element in loop
+            modifier = modifier.getNextSibling();
+        }
+
+        if (!foundScopeModifier) {
+            // prepend string 'package' to mark default scope.
+            // the package scope modifier is placed at the beginning to conform
+            // with modifiers order as suggested in sections 8.1.1, 8.3.1 and 8.4.3 of the JLS
+            collectedModifiers.add(0, "package");
```

Comment:
What if we are in an interface?

Reply:
good catch, I think this should be handled now